src/System.Device.Gpio/System/Device/Gpio/Drivers/LibgpiodCapiDriver.cs

```diff
@@ -0,0 +1,264 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Runtime.InteropServices;
+
+namespace System.Device.Gpio.Drivers
+{
+    public class LibgpiodCapiDriver : GpioDriver
+    {
+        private const string _basePath = "/dev/";
+
+        public string GpioChip { get; set; } = "gpiochip0";
+
+        private SafeChipHandle _chip;
+
+        private Dictionary<int, SafeLineHandle> _pinNumberToSafeLineHandle;
+
+        private timespec _defaultTimeOut = new timespec(IntPtr.Zero, 1_000_000); // one millisecond
+
+        private Dictionary<int, LibgpiodDriverEventHandler> _pinNumberToEventHandler = new Dictionary<int, LibgpiodDriverEventHandler>();
+
+        protected internal override int PinCount
+        {
+            get
+            {
+                return (int)Interop.gpiod_chip_num_lines(_chip);
+            }
+        }
+
+        public LibgpiodCapiDriver()
+        {
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        public LibgpiodCapiDriver(string chip)
+        {
+            GpioChip = chip;
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        private void OpenChip()
+        {
+            _chip = Interop.gpiod_chip_open($"{_basePath}{GpioChip}");
+            if (_chip == null)
+            {
+                throw new GpioChipException($"Chip: {GpioChip} not available, error code: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        private void RequestEvent(int pinNumber, LibgpiodDriverEventHandler eventDescriptor, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.gpiod_chip_get_line(_chip, pinNumber);
+            if (pin != null)
+            {
+                eventDescriptor.PinHandle = pin;
+                if (eventType == PinEventTypes.Rising)
+                {
+                    eventSuccess = Interop.gpiod_line_request_rising_edge_events(pin, $"Listen {pinNumber} for rising edge event");
+                }
+                if (eventType == PinEventTypes.Falling)
```

Comment:
> i [](start = 15, length = 2)

please add "else" here.

Reply:
not really needed, we only support this 2 events and for other case error will be thrown

Reply:
I was trying to point try to avoid testing eventType twice. else will help with that. I'll leave it to you to decide.

---
In reply to: [244898933](https://github.com/dotnet/iot/pull/133#discussion_r244898933) [](ancestors = 244898933)